sympy/printing/printer.py

```diff
@@ -347,39 +347,39 @@ def _as_ordered_terms(self, expr, order=None):
             return expr.as_ordered_terms(order=order)
 
 
-class _PrintFunction:
-    """
-    Function wrapper to replace ``**settings`` in the signature with printer defaults
-    """
-    def __init__(self, f, print_cls: Type[Printer]):
-        # find all the non-setting arguments
-        params = list(inspect.signature(f).parameters.values())
-        assert params.pop(-1).kind == inspect.Parameter.VAR_KEYWORD
-        self.__other_params = params
-
-        self.__print_cls = print_cls
-        update_wrapper(self, f)
-
-    def __repr__(self) -> str:
-        return repr(self.__wrapped__)  # type:ignore
-
-    def __call__(self, *args, **kwargs):
-        return self.__wrapped__(*args, **kwargs)
-
-    @property
-    def __signature__(self) -> inspect.Signature:
-        settings = self.__print_cls._get_initial_settings()
-        return inspect.Signature(
-            parameters=self.__other_params + [
-                inspect.Parameter(k, inspect.Parameter.KEYWORD_ONLY, default=v)
-                for k, v in settings.items()
-            ],
-            return_annotation=self.__wrapped__.__annotations__.get('return', inspect.Signature.empty)  # type:ignore
-        )
-
-
 def print_function(print_cls):
     """ A decorator to replace kwargs with the printer settings in __signature__ """
-    def decorator(f):
-        return _PrintFunction(f, print_cls)
-    return decorator
+    # functools.update_wrapper adds attributes like __doc__ to the class, so
+    # this class needs to be redefined for each decorated function (__doc__
+    # has to go on the class; if it is on the instance it will not work with
+    # help()).
+    class _PrintFunction:
+        """
+        Function wrapper to replace ``**settings`` in the signature with printer defaults
+        """
+        def __init__(self, f):
+            # find all the non-setting arguments
+            params = list(inspect.signature(f).parameters.values())
+            assert params.pop(-1).kind == inspect.Parameter.VAR_KEYWORD
+            self.__other_params = params
+            self.__print_cls = print_cls
+            update_wrapper(type(self), f, updated=())
+
+        def __repr__(self) -> str:
```

Comment:
By the way, I think we shouldn't try to change the repr for this class. I don't think it's a big deal for users to see that the functions are actually wrapper functions. On the other hand, making the repr look like a function can make things very confusing. I was quite confused when debugging this until I noticed this method defined here.

Reply:
No objection from me. I did it this way because it was more useful than the default repr, and easier that deciding on a better one.